components/layout_2020/flow/float.rs

```diff
@@ -70,6 +71,112 @@ impl ContainingBlockPositionInfo {
     }
 }
 
+/// This data strucure is used to try to place non-floating content among float content.
+/// This is used primarily to place replaced content and independent formatting contexts
+/// next to floats, as the specifcation dictates.
+pub(crate) struct PlacementAmongFloats<'a> {
+    /// The [FloatContext] to use for this placement.
+    float_context: &'a FloatContext,
+    /// The current bands we are considering for this placement.
+    current_bands: VecDeque<FloatBand>,
+    /// The size of the object to place.
+    object_size: Vec2<Length>,
+    /// The minimum position in the block direction for the placement. Objects should not
+    /// be placed before this point.
+    ceiling: Length,
+}
+
+impl<'a> PlacementAmongFloats<'a> {
+    pub(crate) fn new(
+        float_context: &'a FloatContext,
+        ceiling: Length,
+        object_size: Vec2<Length>,
+    ) -> Self {
+        let current_bands = VecDeque::from([float_context.bands.find(ceiling).unwrap()]);
+        PlacementAmongFloats {
+            float_context,
+            current_bands,
+            object_size,
+            ceiling,
+        }
+    }
+
+    fn top_of_placement_for_current_bands(&self) -> Length {
+        self.ceiling.max(self.current_bands.front().unwrap().top)
+    }
+
+    fn current_bands_height(&self) -> Length {
+        assert!(self.current_bands.len() > 0);
+        self.current_bands.back().unwrap().top - self.top_of_placement_for_current_bands()
+    }
+
+    fn accumulate_enough_bands_for_block_size(&mut self) {
+        while self.current_bands_height() < self.object_size.block {
+            assert!(self.current_bands.len() > 0);
+            let next_band = self
+                .float_context
+                .bands
+                .find_next(self.current_bands.back().unwrap().top)
+                .unwrap();
+            self.current_bands.push_back(next_band);
+        }
+    }
+
+    fn calculate_viable_inline_space(&self) -> (Length, Length) {
+        let mut max_inline_start = self.float_context.containing_block_info.inline_start;
+        let mut min_inline_end = self.float_context.containing_block_info.inline_end;
+        for band in self.current_bands.iter() {
```

Comment:
I think this should avoid the last band. The last band is only there to know the height of the bands, but we shouldn't actually consider it.

```html
<!DOCTYPE html>
<div style="display: flow-root; border: solid; width: 200px">
  <div style="float: left; height: 100px; width: 150px; background: cyan"></div>
  <div style="float: left; height: 100px; width: 175px; background: lime"></div>
  <div style="display: flow-root; background: magenta; width: 50px; height: 50px"></div>
</div>
```

Even though the magenta element would fit next to the cyan float, due to iterating here we think that we can only fit 25px, so the magenta element goes below the floats. This doesn't match other browsers.

Reply:
Oh, nice catch! I've fixed this and a few more tests are now passing.